### DISCUSSION TITLE
Level Set simulations for 3D geometries

### DISCUSSION THREAD

**Discussion by Rave169** [usnistgov/fipy#901] [Q&A]
Hi FiPy Team,

I am trying to do some level set simulations for 3d geometries on a windows machine with scikit-fmm library installed. When I try to create a 3d geometry, I am getting an error as '3D meshes not yet implemented'. Below is the simple code for generating 3D structures.

Is scikit-fmm capable to do 3D level set simulations or LSMLIB package is necessary?

```
from fipy import Grid3D, DistanceVariable
from fipy.tools import numerix, serialComm

Lx = 1.0
Ly = 1.0
Lz = 1.0
dx = Lx/128
dy = Ly/128
dz = Lz/128
nx = int(Lx/dx)
ny = int(Ly/dy)
nz = int(Lz/dz)

shape = (nx, ny, nz)

vel = 1
cfl = 0.1
timeStep = cfl*dx/vel
steps = 200
    
trenchTop = 0.25*Ly
trenchBot = 0.01*Ly
trenchLeft = 0.25*Lx
trenchRight = 0.75*Lx 

aspectRatio = (trenchRight-trenchLeft)/(trenchTop-trenchBot)

mesh = Grid3D(dx=dx, dy=dy, dz=dz, nx=nx, ny=ny, nz=nz, communicator=serialComm)
x, y, z = mesh.cellCenters

distanceVar = DistanceVariable(name = 'levelset variable', mesh=mesh, value=-1., hasOld=1)
distanceVar.setValue(1., where=(y>=trenchTop))
distanceVar.setValue(1., where=((y>trenchBot) & ((x>trenchLeft) & (x<trenchRight))))
distanceVar.calcDistanceFunction()
```

**Comment by wd15** [comment]
Both LSMLIB and scikit-fmm support 3D meshes. It's the FiPy distance variable that does not support 3D. It is relatively easy to use Scikit-fmm independently, in the sense of using it's functions directly and feeding the result into to a FiPy cell variable that's used in in FiPy to calculate diffusion coefficients or velocities. If you need more complex calculations (like surface surfactants or surface reactions) then you'll obviously need to implement those in 3D.
